cli/kubernetes/interactive_options_reader.go

```diff
@@ -21,10 +21,16 @@ func NewInteractiveOptionsReader(stdout io.Writer, stdin io.Reader) InteractiveO
 	return InteractiveOptionsReader{in: stdin, out: stdout}
 }
 
-// Read aks the user what value should the given InstallationOption have and
+// Read asks the user what value should the given InstallationOption have and
 // returns that value validated and converted to the appropriate type as defined
 // by the Type field of the InstallationOption.
-func (reader InteractiveOptionsReader) Read(option InstallationOption) (interface{}, error) {
+func (reader InteractiveOptionsReader) Read(option *InstallationOption) error {
+	// Ignore anything which is already valid, be it default, or
+	// user-specified (cli option, etc.)
+	if option.Valid {
+		return nil
```

Comment:
I think the common case for cli's is that when you are in interactive mode and a default value exists something like this is printed:

```
$ system_domain [172.19.0.2.nip.io] :  # (Hit enter if default is fine)
```
and you can simply accept the default by hitting "Enter". This UX might be better because:

1. that's what people are used to
2. Allows you to override defaults in interactive mode
3. Doesn't need the `Valid` flag (which could be used then solely for actual Validation purposes)

Every other reader (other than the Interactive one) would implicitly/silently use the default value when another one is not available. 

If you agree with the above, then we don't need a DefaultOptionsReader but simply a `Default()` method which would try a hardcoded default and the dynamic default (the function). The dynamic default function should probably cache its result on the `Default` attribute so that its faster the next time the default value is requested.

Reply:
I should possibly call the `Valid` flag: `UserSpecified`. In the current code this is nearer to how it is used. And in your proposal it would be exactly how it would be used. IOW interactive mode asks for any option but those already specified by the user.

My understanding of you idea for the `Default()` method is that it would be used by the interactive reader to show and set the value. Should also have cli option `--non-interactive`, (`-n`) where the interactive would simply accept all the defaults.

With respect to caching, each option/default is used only once at the moment. There is only the `install` command. And even in case of multiple commands the options (declarations) would not be not shared between them in the current setup.

Validation purposes: Would not need a `Valid` flag as a failed validation would terminate early, printing an error, not set a flag to abort late.

Reply:
- +1 for the `--non-interactive`
- Maybe it's nit picking at the moment but we should be able to call `Default()` simply to check what the default value is anywhere in the code without worrying that this may trigger some slow `kubectl` command (e.g. find the external ip of the traefik service). Currently it wouldn't make any difference that's true but when it becomes important we may forget it needs refactoring. Feel free to postpone this for when it becomes an issue.
- Regarding Valid flag I was thinking that in the case of non-interactive readers (e.g. yaml configuration) we may want to check the validity of all options first and then print all the errors at once so the user doesn't have to call the installer again and again until she discovers all the problems.

Reply:
Feel free to postpone anything that can be postponed for later (e.g. `--non-interactive` flag, caching etc)

Reply:
Addressed with commit f1027e9040f094d29697aa5e7a16c88ccd9e55cc

Reply:
I hate it when github does not update properly, hiding new comments from me

> Maybe it's nit picking at the moment but we should be able to call Default() simply to check what the default value is anywhere in the code without worrying that this may trigger some slow kubectl command (e.g. find the external ip of the traefik service). Currently it wouldn't make any difference that's true but when it becomes important we may forget it needs refactoring. Feel free to postpone this for when it becomes an issue.

This makes sense. I would place that into a separate field though, `CachedDefault` or some such. Would be needed because currently the func has precedence over `Default`, so caching into that field would simply not work.

> Regarding Valid flag I was thinking that in the case of non-interactive readers (e.g. yaml configuration) we may want to check the validity of all options first and then print all the errors at once so the user doesn't have to call the installer again and again until she discovers all the problems.

True. So, interactive reader does validation as user enters values, and a separate validation phase at the end, to catch everything from yaml files, options, env.variables, and so on, after everything is resolved to final values.